---
title: "GSE_networks"
author: "Theo Killian"
date: "7/30/2019"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, results = 'asis'}
# suppressPackageStartupMessages()
knitr::opts_chunk$set(collapse=TRUE, comment="#>", warning=FALSE, message=FALSE)
```

# Enhancing gene set enrichment using networks

This `R` pipeline implements a network analysis on the enriched results matrix
generated from a `DESeq2` differential expression analysis derived from a RNAseq
count matrix from mouse or human tissue. Automated community detection of
differentially regulated gene sets is performed, categorizating genes into
singletons and clusters of different size ranges. Automated label generation is
allowed to associate these clusters with biological themes or processes of which
the member gene sets are part of (such as `KEGG`, `Reactome`, and `Biocarta`
pathways), based on a Jaccard index of a given threshold. 

## Preparations

The packages required for this workflow provide plotting functions (`ggplot2`
and relatives), network functions `igraph` and `GGally`, text analytics
functions (`wordcloud`, etc.) and gene expression analysis functions `DESeq2`,
`limma`, and `org.Mm.eg.db` (or `org.Hs.eg.db`).

```{r load_library}
library("RColorBrewer")
library("ggplot2")
library("gplots")
library("cowplot")
library("ggrepel")
library("reshape2")
library("knitr")
library("kableExtra")
library("igraph")
library("GGally")
library("DESeq2")
library("limma")
library("wordcloud")
library("tm")
library("SnowballC")
library("intergraph")
library("DT")
# depending on the source of the data, comment out and use the appropriate
# library for mouse or human
# library("org.Hs.eg.db") # for human
library("org.Mm.eg.db") # for mouse
```

## Load data

This pathway analysis pipeline accepts count matrix files with *Ensembl ID* row
names and performs a differential expression (*DE*) analysis via `DESeq2`.

```{r load_data}
## read count matrix file
countdata <- read.delim("count_matrix", row.names=1)
head(countdata)
```

## DESeq2 analysis

```{r assign_condition}
## Assign condition vector to 'coldata' dataframe 
condition <- factor(c(rep("ctl", 4), rep("hr2", 4), rep("hr4", 4)))
coldata <- data.frame(row.names=colnames(countdata), condition)
head(coldata)
```

```{r DESeq2}
## Instantiate the DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData=countdata,
                              colData=coldata,
                              design=~condition)

## filtering out genes with counts of all 0 for all samples 
keep <- rowSums(counts(dds)) > 0

## Run DESeq2
dds <- DESeq(dds[keep, ])
```

We obtain the results of the `DESeq2` analysis, with significance cutoff used
for optimizing the independent filtering (i.e. *alpha*) set at 0.05. 

```{r res}
## get results of the comparison between "ctl" vs "hr2"
res <- results(dds, contrast=c("condition", "ctl", "hr2"), alpha=0.05)
```

## Mapping Ensembl IDs to ENTREZ IDs

This step maps the *Ensembl IDs* to *ENTREZ IDs* using the popular `org.Hs.eg.db`
and `org.Mm.eg.db` packages based on the `UCSC` annotation database and keep
only genes with a unique mapping.

```{r mapping, warning=FALSE, message=FALSE}
# uncomment lines depending on appropriate organism

## match with human
# res$entrezgene <- unname(mapIds(org.Hs.eg.db, keys = rownames(res),
#                                column = "ENTREZID", keytype = "ENSEMBL"))

# match with mouse
res$entrezgene <- unname(mapIds(org.Mm.eg.db, keys = rownames(res),
                               column = "ENTREZID", keytype = "ENSEMBL"))
res <- subset(res, subset = !is.na(res$entrezgene) & !is.na(res$stat))
res <- res[-which(duplicated(res$entrezgene)), ]
```

## Gene set enrichment analysis (GSEA)

This step searches for Mouse and human orthologs of the `MSigDB` containing
`KEGG`, `Reactome`, and `Biocarta` pathways from the `MSigDB` gene set library
*C2*. The following chunk guarantees that the gene set library list object is
called gset.

```{r GSEA}
# uncomment lines depending on appropriate organism, see:
# http://bioinf.wehi.edu.au/software/MSigDB/
# currently only libraries for mouse or human are available

# for human
# url <- "http://bioinf.wehi.edu.au/software/MSigDB/human_c2_v5p2.rdata"

# for mouse
url <- "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5p2.rdata"
temp.space <- new.env()
bar <- load(url(url), temp.space)
gset <- get(bar, temp.space)
rm(temp.space)
gs.libs <- sapply(names(gset), function(x) strsplit(x, "_")[[1]][1])
gset <- gset[which(gs.libs %in% c("KEGG", "REACTOME", "BIOCARTA"))]
```

Competitive gene set enrichment analysis is performed using the function
`camera()` from the `limma` package. We include unidirectional and bidirectional 
enrichment by using both the test statistics ("up" or "down") and its modulus
("mixed") for gene set testing. We limit the following network analysis to gene
sets with a FDR < 0.05.

```{r camera}
idx <- ids2indices(gene.sets = gset, identifiers = res$entrezgene)
dat <- cameraPR(res$stat, idx, sort = FALSE)
dat$PValue.Mixed <- cameraPR(abs(res$stat), idx, sort = FALSE)$PValue
dat$FDR.Mixed <- p.adjust(dat$PValue.Mixed, method = "BH")
dat$name <- rownames(dat)

dat$Direction <- as.character(dat$Direction)
dat$Direction[dat$FDR > 0.05] <- "Mixed"
dat$Direction[dat$Direction == "Mixed" & dat$FDR.Mixed > 0.05] = "NOT"
dat$Direction <- factor(dat$Direction, levels=c("NOT", "Up", "Down", "Mixed"))

idx <- which(dat$Direction == "Mixed")
if(length(idx) > 0) dat$FDR[idx] = dat$FDR.Mixed[idx]
dat <- dat[, -grep("\\.Mixed", names(dat))]
dat <- dat[dat$Direction != "NOT", ]
dat$Direction <- factor(dat$Direction, levels=c("Up", "Down", "Mixed"))
```

Starting from all gene sets, a subset are found to be differentially regulated.
Many of them are expected to describe similar processes and to be highly
correlated.

## Network construction

We construct a gene set network based on the proportion of common genes as the
inverse distance measure. The nodes are gene sets which are connected by edges
if the Jaccard index is larger than a preset threshold, J > 0.2. While this
threshold is somewhat arbitrary it has proven to be a reasonable one in many
projects. Nevertheless, it is strongly recommended to investigate its effect on
the quality of the results.

```{r net_construction}
# only keep gene sets present in the data
id.keep <- which(names(gset) %in% dat$name)
gset <- gset[id.keep]

# adjacency matrix
m.adj <- sapply(gset, function(x)
  sapply(gset, function(y)
    length(intersect(unlist(x), unlist(y) ))
    )
  )
diag(m.adj) = 0

# Jaccard index matrix
NGenes <- sapply(gset, length)
m.union <- outer(NGenes, NGenes, "+") - m.adj
m.jacc <- m.adj / m.union
```

The Jaccard matrix, or adjacency matrix, can be conveniently used to construct a
network object using the function `igraph::graph_from_adjacency_matrix()`. In
this example geneset, similarity is measured using all member genes irrespective
of whether they were detected and present in the data. Alternatively, one could
include only genes present in the data depending on whether the current data
seem more relevant and trustworthy or the prior information given by the geneset
definition. Graphical display is achieved here using `ggnet::ggnet2()`

```{r jaccard}
# choose node colors
palette <- brewer.pal(9, "Set1")[c(1,2,9)]
names(palette) <- c("Up", "Down", "Mixed")

# apply cutoff to Jaccard matrix
m.adj1 <- m.adj * (m.jacc > 0.2)

# construct network object
net <- graph_from_adjacency_matrix(m.adj1, "upper",
                                  diag = FALSE, weighted = TRUE)

# add vertex features
V(net)$size <- dat$NGenes
V(net)$color <- palette[dat$Direction]
V(net)$Direction <- as.character(dat$Direction)

# plot
ggnet2(net, size = 2, color = "Direction", palette = palette, edge.size = 1,
       edge.color = "#99CC33")
```

## Network modifications

In the following, components of the network for which network analysis does not
improve interpretability are identified and put to aside. This includes
singletons, i.e., genesets not connected to any other geneset, and doublets,
also termed binary systems or dumbbells, i.e., pairs of genesets connected with
each other but isolated from the rest.

## Identify singletons

This step identifies "singletons" which are genesets without sufficient overlap
with any other geneset. A table is shown below of the genes of the singleton
grouping.

```{r singletons}
singletons <- which(igraph::degree(net) == 0)
net1 <- delete_vertices(net, singletons)
in.single <- which(dat$name %in% V(net)$name[singletons])
tab <- dat[in.single, ]
tab$FDR <- signif(tab$FDR, 2)
tab$name <- gsub("_", " ", tab$name)
# tab <- kable(tab[,c("name", "NGenes", "Direction", "FDR")],
#               row.names = FALSE, format = "latex",
#               caption = "List of all singletons, i.e., genesets without
#               sufficient overlap with any other geneset.")
# kable_styling(tab, latex_options = "scale_down", font_size = 8)
DT::datatable(tab)
```

Here, singletons were identified and excluded from further analysis
*(Table 1)*. It is important to note that these genesets, while down-prioritized
for the time being, may still be worthwhile investigating later.

```{r singletons_plot}
ggnet2(net1, size = "size", max_size = 4, color = palette[V(net1)$Direction],
       size.cut = 4, edge.size = 1, edge.color = "#99CC33")
```

## Identify binary systems (2 sets)

Next we also want to separate clusters with less than 3 gene sets. To do so, we
separate disjoint subnets as individual objects, count their members, and delete
all vertices belonging to clusters of size smaller than 3. A table is shown
below of the genes of this binary grouping.

```{r binary}
clu1 <- igraph::components(net1)
clu.lt3 <- which(sizes(clu1) < 3)
v.clu.lt3 <- which(clu1$membership %in% clu.lt3)
net2 <- delete_vertices(net1, v.clu.lt3)
clu2 <- igraph::components(net2)
in.clu.lt3 <- which(dat$name %in% V(net1)$name[v.clu.lt3])
tab <- dat[in.clu.lt3, ]
tab$FDR <- signif(tab$FDR,2)
cludp <- clu1$membership[v.clu.lt3]
cludp <- data.frame(name = names(cludp), id = as.numeric(cludp))
tab <- merge(tab,cludp)
tab$name <- gsub("_", " ", tab$name)
# tab <- kable(tab[order(tab$id), c("id", "name", "NGenes", "Direction", "FDR")],
#              row.names=FALSE, format = "latex",
#              caption = "List of binary clusters as indicated by the id column.")
# kable_styling(tab, latex_options = "scale_down", font_size = 8)
DT::datatable(tab)
```

In *Table 2*, consecutively listed gene sets with the same id belong to the same
binary cluster. Often these are gene sets from different libraries describing
the same biological process or phenotype. Here, binary clusters are identified,
for which network analysis may or may not be useful.

```{r binary_plot}
set.seed(16)
nodecol <- colorRampPalette(brewer.pal(9,
                                       "Set1")[sample(9)])(max(clu2$membership))
ggnet2(net2, size = "size", max_size = 4, color = nodecol[clu2$membership],
       size.cut = 4, edge.size = 1, edge.color = "grey")
```

Without singletons and binary clusters, we are left with larger disjoint subnets

## Detect communities (sub-networks)

The larger disjoint clusters may consist of so-called communities, i.e.,
sub-networks of highly inter-connected nodes that stick together by only one or
a few edges. We are using the popular edge betweenness property to identify
these community-connecting edges and remove them in order to split large
clusters into smaller ones.

```{r sub_networks}
net2 <- delete_edge_attr(net2, "weight")
clu3 <- cluster_edge_betweenness(net2)

# delete edges between communities
net3 <- delete_edges(net2, which(as.vector(crossing(clu3, net2))))

# remove clusters of size <3
small_cluster_ids <- which(sizes(clu3) < 3)
small_cl_v <- which(clu3$membership %in% small_cluster_ids)
net3 <- delete_vertices(net3, small_cl_v)

clu3 <- igraph::components(net3)
nodecol <- c(brewer.pal(9, "Paired"), brewer.pal(9, "Set3") )
nodecol <- colorRampPalette(nodecol)(max(clu3$membership))

ggnet2(net3, size = 0, color = nodecol[clu3$membership],
       edge.size = 1.0, edge.color = "grey") +
       geom_point(size = 2, color = "black") +
       geom_point(aes(color = color), size = 1)
```

## Automatic annotation of gene set clusters

In analogy to the popular interactive network visualization tool *cytoscape*,
we attempt to generate automatic labels for gene set clusters. Gene set names
are split into individual words and counted within each cluster. The four most
frequent terms occurring at least twice are used as labels. The function
`clust_head()` is defined for this purpose and contains an exclusion list of
words not used.

```{r auto_annotation}
t.rW <- c("cell", "process", "regulation", "negative", "positive", "signaling",
          "response", "stimulus", "signal", "activity", "protein", "involved",
          "component", "level", "effector", "event", "projection", "organismal",
          "cellular", "modification", "pathway", "mediated", "dependent",
          "organization", "group", "target", "biocarta", "kegg", "reactome")

clust_head <- function(x){
  txt <- unlist(strsplit(x, "_"))
  txt <- Corpus(VectorSource(txt))
  txt <- tm_map(txt, PlainTextDocument)
  txt <- tm_map(txt, removePunctuation)
  txt <- tm_map(txt, removeNumbers)
  txt <- tm_map(txt, content_transformer(tolower))
  txt <- tm_map(txt, removeWords, c(t.rW, stopwords("english")))
  tdm <- TermDocumentMatrix(txt)
  m <- as.matrix(tdm)
  word_freqs <- sort(rowSums(m), decreasing=TRUE)
  word_freqs <- word_freqs[word_freqs>1]
  word_freqs <- paste(names(word_freqs)[1:4], collapse=" ")
  gsub("[[:space:]]?NA[[:space:]]?", "", word_freqs)
}
```

## Lattice of annotated networks

There are many possibilities to visualize geneset clusters and often a
compromise between information content and crowding has to be found. Here, we
are producing a lattice of network plots, one for each sub-net, with the
automatic annotation as title *(Figure 5)*. We begin by generating the cluster
titles using the `clust_head()` function followed by cleaning up and ordering by
cluster size.

```{r network_lattice, warning=FALSE, message=FALSE}
clust <- data.frame(cl = clu3$membership)
rownames(clust) <- names(V(net3))

# generate cluster titles
cl3.lab.txt <- as.character(tapply(rownames(clust), clust$cl, clust_head))

# remove NAs
cl3.lab.txt <- gsub("[[:space:]]?NA[[:space:]]?", "", cl3.lab.txt)
clu3 <- igraph::components(net3)
clu.order <- order(clu3$csize, decreasing = TRUE)
clu3$mem <- match(clu3$membership, clu.order)
```

Then we generate a list of ggplot objects, one for each cluster or sub-net. For
smaller sub-nets, the nodes are labelled with the first 4 words of their names;
the first word was removed before as it is usually the name of the geneset
library. For larger sub-nets, this is not feasible without overprinting. Titles
are missing if none of the words from the geneset names occurred more than once.

```{r list_of_ggplots}
# generate a list of ggplots
g <- list(max(clu3$membership))
set.seed(7042016)
for (ii in 1:max(clu3$membership)) {
  subgf <- induced_subgraph(net3, which(clu3$mem == ii))
  
  # generate titles with one optional line break
  title <- substr(toupper(cl3.lab.txt[clu.order][ii]), 1, 60)
  if (nchar(title) > 25) {
    title <- sub("(^.{10,30})[[:space:]]","\\1\\\n", title)
  }
  # generate node labels using word 2-5 of the geneset name
  v.label <- names(V(subgf))
  v.label <- lapply(v.label, function(x) strsplit(x, "_")[[1]])
  v.label <- sapply(v.label, function(x) paste(x[2:min(5, length(x))],
                                              collapse = "_"))
  # clean up geneset names
  v.label <-  gsub("_PATHWAY","", v.label)
  v.label <-  gsub("_SIGNALING", "", v.label)
  
  # introduce line breaks
  v.label <-  gsub("_","\n", v.label)
  
  # remove node labels for large clusters
  if (length(v.label) > 5) v.label = rep(NA, length(v.label))
  g[[ii]] = ggnet2(subgf, edge.size = 1, edge.color = "#99CC33",
                     label = FALSE, size=V(subgf)$size, max_size = 3,
                     size.cut = 4, color = palette[V(subgf)$Direction]) +
    theme(legend.position="none", plot.title = element_text(size=6),
           panel.grid = element_blank()) +
    geom_label_repel(label = v.label, size=1.2,
                       box.padding = 0.1, label.padding = 0.1) +
    ggtitle(title)
  }
```

```{r plot_nets, warning=FALSE, message=FALSE}
nr.cols <- min(4,max(clu3$membership))
nr.rows <- ceiling(max(clu3$membership) / nr.cols)
width <- sapply(g, function(x) nrow(x$data))
grid.arrange <- getFromNamespace("grid.arrange", asNamespace("gridExtra"))
grid.arrange(grobs = g[seq(16)], ncol = nr.cols)
```

We have presented an automated workflow based on a small number of `R` packages
for prioritization and visualization of gene set analysis results using
networks, which we call *RICHNET*. We demonstrated how community detection
facilitates categorization of differentially regulated gene sets into singletons
and clusters of different size ranges. Automated label generation allowed to
associate these clusters with biological themes or processes of which the member
gene sets are part of.

```{r sessionInfo}
sessionInfo()
```
